File: src/3607.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class DSU:
    def __init__(self, size: int) -> None:
        self.parent = list(range(size))

    def find(self, x: int) -> int:
        retVal = 0

        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        retVal = self.parent[x]

        return retVal

    def join(self, u: int, v: int) -> None:
        self.parent[self.find(v)] = self.find(u)


class Solution:
    def processQueries(self, c: int, connections: List[List[int]], queries: List[List[int]]) -> List[int]:
        retVal = []

        dsu = DSU(c + 1)
        for p in connections:
            dsu.join(p[0], p[1])

        online = [True] * (c + 1)
        offline_counts = [0] * (c + 1)
        for q in queries:
            op, x = q[0], q[1]
            if op == 2:
                online[x] = False
                offline_counts[x] += 1

        minimum_online_stations = {}
        for i in range(1, c + 1):
            root = dsu.find(i)
            if root not in minimum_online_stations:
                minimum_online_stations[root] = -1

            station = minimum_online_stations[root]
            if online[i]:
                if station == -1 or station > i:
                    minimum_online_stations[root] = i

        for i in range(len(queries) - 1, -1, -1):
            op = queries[i][0]
            x = queries[i][1]

            root = dsu.find(x)
            station = minimum_online_stations[root]

            if op == 1:
                if online[x]:
                    retVal.append(x)
                else:
                    retVal.append(station)

            if op == 2:
                if offline_counts[x] > 1:
                    offline_counts[x] -= 1
                else:
                    online[x] = True
                    if station == -1 or station > x:
                        minimum_online_stations[root] = x

        retVal = retVal[::-1]

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for c, connections, queries in [
            [5, [[1, 2], [2, 3], [3, 4], [4, 5]], [[1, 3], [2, 1], [1, 1], [2, 2], [1, 2]]],
            [3, [], [[1, 1], [2, 1], [1, 1]]]
        ]:
            # /* Example
            #  *  Input: c = 5, connections = [[1,2],[2,3],[3,4],[4,5]], queries = [[1,3],[2,1],[1,1],[2,2],[1,2]]
            #  *  Output: [3,2,3]
            #  *
            #  *  Input: c = 3, connections = [], queries = [[1,1],[2,1],[1,1]]
            #  *  Output: [1,-1]
            #  */
            logging.info("Input: c = %s, connections = %s, queries = %s", c, connections, queries)

            retVal = pSolution.processQueries(c, connections, queries)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)

File: src/3607.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <unordered_map>
#include <vector>

using namespace std;

class DSU {
   public:
    vector<int> parent;

    DSU(int size) {
        parent.resize(size);
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int x) {
        int retVal = x;

        if (parent[x] != x) {
            parent[x] = find(parent[x]);
            retVal = parent[x];
        }

        return retVal;
    }
    void join(int u, int v) {
        //
        parent[find(v)] = find(u);
    }
};
class Solution {
   public:
    vector<int> processQueries(int c, vector<vector<int>>& connections, vector<vector<int>>& queries) {
        vector<int> retVal;

        DSU dsu(c + 1);
        for (auto& p : connections) {
            dsu.join(p[0], p[1]);
        }

        vector<bool> online(c + 1, true);
        vector<int> offlineCounts(c + 1, 0);
        for (auto& q : queries) {
            int op = q[0], x = q[1];
            if (op == 2) {
                online[x] = false;
                offlineCounts[x]++;
            }
        }

        unordered_map<int, int> minimumOnlineStations;
        for (int i = 1; i <= c; i++) {
            int root = dsu.find(i);
            if (!minimumOnlineStations.count(root)) {
                minimumOnlineStations[root] = -1;
            }

            int station = minimumOnlineStations[root];
            if (online[i]) {
                if (station == -1 || station > i) {
                    minimumOnlineStations[root] = i;
                }
            }
        }

        for (int i = (int)queries.size() - 1; i >= 0; i--) {
            int op = queries[i][0], x = queries[i][1];
            int root = dsu.find(x);
            int station = minimumOnlineStations[root];

            if (op == 1) {
                if (online[x]) {
                    retVal.push_back(x);
                } else {
                    retVal.push_back(station);
                }
            }

            if (op == 2) {
                if (offlineCounts[x] > 1) {
                    offlineCounts[x]--;
                } else {
                    online[x] = true;
                    if (station == -1 || station > x) {
                        minimumOnlineStations[root] = x;
                    }
                }
            }
        }

        reverse(retVal.begin(), retVal.end());

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        int c;
        vector<vector<int>> connections;
        vector<vector<int>> queries;
    };
    vector<subject> testData{{5, {{1, 2}, {2, 3}, {3, 4}, {4, 5}}, {{1, 3}, {2, 1}, {1, 1}, {2, 2}, {1, 2}}},
                             {3, {}, {{1, 1}, {2, 1}, {1, 1}}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: c = 5, connections = [[1,2],[2,3],[3,4],[4,5]], queries = [[1,3],[2,1],[1,1],[2,2],[1,2]]
     *  Output: [3,2,3]
     *
     *  Input: c = 3, connections = [], queries = [[1,1],[2,1],[1,1]]
     *  Output: [1,-1]
     */

    Solution cSolution;
    vector<int> answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: c = " << testData[i].c << ", connections = [";
        for (long unsigned int j = 0; j < testData[i].connections.size(); ++j) {
            cout << (j == 0 ? "" : ",") << "[";
            for (long unsigned int k = 0; k < testData[i].connections[j].size(); ++k) {
                cout << (k == 0 ? "" : ",") << testData[i].connections[j][k];
            }
            cout << "]";
        }
        cout << "], queries = [";
        for (long unsigned int j = 0; j < testData[i].queries.size(); ++j) {
            cout << (j == 0 ? "" : ",") << "[";
            for (long unsigned int k = 0; k < testData[i].queries[j].size(); ++k) {
                cout << (k == 0 ? "" : ",") << testData[i].queries[j][k];
            }
            cout << "]";
        }
        cout << "]\n";

        answer = cSolution.processQueries(testData[i].c, testData[i].connections, testData[i].queries);
        cout << "Output: [";
        for (long unsigned int j = 0; j < answer.size(); ++j) {
            cout << (j == 0 ? "" : ",") << answer[j];
        }
        cout << "]\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}
